### ISSUE TITLE
Improving the collaborative process between MDG and community

### ISSUE THREAD

**Issue by KoenLav** [meteor/meteor#10477]
❗️ **Bug**
I think I found a bug in the way the development of Meteor is currently organized and would like to work together with the community to resolve this.

Instead of hijacking the 1.8.1 release thread (sawry), I'd like to get a constructive discussion going here with the end goal of establishing better communication between MDG and the community.

**The Issue**
The **resources** which MDG currently **commits** to Meteor, a rather **large** open-source framework, combined with the (lack of) **communication** around this, causes **uncertainty** which leaves the **community** with insufficient **confidence** in the future of Meteor.

**Key terms**
- Resources: Meteor is too big to run by a one-man team;
- Communication: Lack of communication causes uncertainty.

**Comment by smeijer** [comment]
I'm supporting a topic like this for the full 100%. Like I've mentioned in the release branch, part of me feels wrong about "complaining", as it's a free product. And we often have the (unspoken) rule of not to complain about something that's free. Or that we should fix it by contributing or forking ourselves, as that's the spirit of open source.

That being said, I also believe that a company that got a **32 million USD investment**, should be able to handle the community that made this investment possible. I strongly believe that it's the community that makes investments like this possible. Even though the brains behind MDG deserve some serious respect.

2012 - https://blog.meteor.com/meteors-new-11-2-million-development-budget-7370586949e7
2015 - https://blog.meteor.com/announcing-our-20m-series-b-funding-49dcfd3c3c6f

The 2012 blog, contains the following quote of Geoff, CEO of MDG:

> Our agenda for the next few years is as follows:
> - **Support the Meteor community**. This includes everything from publishing books and organizing conferences, to **being responsive to bug reports and pull requests**, to finally making some cool tshirts.

The 2015 blog, contains the following quote of Geoff, CEO of MDG:

> Our continuing commitment to you is to ensure that we deliver on 3 things:
> - **Meteor**. ... 
> - **Galaxy**. ...
> - **Community**. The software is only half of the picture. The Meteor community is the other, equally important half. We would be nowhere without your support in building, evangelizing, and extending the platform. Meteor Development Group exists to serve and support you, so let us know how we can help, and we will listen and do our very best.

So, after receiving a 20 milion dollar investment, the community was "half of the picture". I can definitly agree on that. @gschmidt, you're welcome!

To get the timing right, the 20 milion USD investment was May 19, 2015. Then in November 18, 2015 Geoff published a post on the meteor forms in response to all that was happening around the Blaze (communication) fiasco.

2015 - https://forums.meteor.com/t/next-steps-on-blaze-and-the-view-layer/13561

You can stop reading after the first paragraph, let me quote:

> Hi everyone,
> 
> For those of you I haven’t met, I’m Geoff Schmidt, one of the original Meteor core developers and CEO of MDG. With Galaxy in the field now and doing great, I’ve been able to clear out a big chunk of my schedule and **you can expect to see me much more active on the forums.**

I've done [some analyses before](https://forums.meteor.com/t/meteor-needs-long-term-support/37262/32?u=smeijer), and came to the conclusion that his activity didn't really grow after that statement:

```
              Geoff          Sashko
Joined:       Feb 27, '15    Feb 24, '15
Last Post:    Feb 22, '16    May 1,  '18 
Seen          May 10, '16    Jul 12, '18

STATS
visited:         86 days       807 days
read time:        7 hours      240 hours

topics viewd:    51          5 900
posts read:     638         35 100
topics created:   6             71 
posts created:   22          3 000
```

A comparison I made on 5 October 2018, between Geoff's activity and Sashko's activity. At this moment, I cannot create new stats as [his profile](https://forums.meteor.com/u/gschmidt) has been made private.

To quote myself from that topic (link above):

> He hasn’t been here for over 2 years. Only three months after the post mentioned above, right in the rural times of Meteor. He decided not to come back to the forums and listen to or get involved with the community.

---

Don't see the part above this line as crying or complaining. I believe it's important to know your history, to be able to learn and move on.

Point being, I've lost faith in changes. The year of the 20 milion USD investment, is the year they lost the connection with the community. This is confirmed by this SO trend line:

[![Meteor Trendline](https://user-images.githubusercontent.com/1196524/53635496-9c915880-3c1d-11e9-9cc5-d9cf87efe258.png)](https://insights.stackoverflow.com/trends?tags=meteor)

source: https://insights.stackoverflow.com/trends?tags=meteor

We're dealing here with a CEO that promises to get involved, but doesn't. Leaving the impression that he took a nice share of the 32 milion, and just doesn't care any more.

I wouldn't be surpised if Apollo would get a nice funding any time soon.

---

**TLDR**

I lost faith in changing this company, and try make it listen to it's community. They have/had a few awesome developers there (`@stubailo`, `@benjamn`), but I'm not a fan of management, so to say.

My hope is in creating a fork. There are some great ideas out there that can be used for an initial roadmap; see https://forums.meteor.com/t/thingking-about-meteor-2-0/46022

Perhaps some great contriburos like @zodern, @yorrd, @StorytellerCZ, @KoenLav have thoughts about this, and can chime in it. (not in a specific order, and sorry if I missed your name).

**Comment by morloy** [comment]
I see two options for making Meteor the great thing it really can be:
1. Adding official maintainers from the community
2. Create a community-maintained fork

For MDG, I have the feeling that they are too much focused on Apollo. It seems to be a great project and a good path for MDG. But the current state of Meteor misses many exciting opportunities.

**Comment by benjamn** [comment]
Can I ask a concrete question of this group?

What specifically matters to you about the final official Meteor 1.8.1 release that isn't satisfied by the beta releases?

The long release cycle exists so that folks who need the new stuff can begin using it sooner, and those who can wait for the official release will (eventually) benefit from all the testing and validation that went into the beta/RC process.

From my perspective, if you're a Meteor developer in 2019 and you don't know about the beta/RC release cycle, you're probably not actively invested in Meteor development, or in Meteor's future. And that's fine, but it means you're enjoying the benefits of free software without helping to ensure its longevity. And if the relative infrequency of official releases bothers you, you've probably already figured out that the beta/RC cycle is the place to be.

I've tried in the past to make predictions about when official releases will happen, and it just doesn't work. Unexpected contingencies like Cordova bugs emerge at surprising times, and demand my/our attention. I'm sorry that I've ignored requests for new predictions, but please understand: my predictions will always be pretty vague, because I don't fully understand what you're (collectively) looking for in an official release.

If you have a specific feature or bug fix that you'd like to see released officially, and you think it's ready to ship, I would really appreciate that kind of feedback. Obviously those requests need to be balanced against any bugs that other folks are expecting to see fixed, but concrete reasons to cut an official release would help a lot. Right now the 1.8.1 release is in a sort of limbo where I'm waiting to be sure the update process will be smooth, and passive developers will be glad they updated. Are we there yet? I genuinely need your help to make that call.

**Comment by KoenLav** [comment]
@benjamn wow, claiming that the people mentioned in this thread aren't actively involved. 

There's a lot of things in your answer I'd like to reply to. But let me first reply to this; if MDG needed the help to decide, MDG should have asked. 

Communication.

Even without MDG asking, people told their stories, for instance the Cordova/HCP changes.

I'll make sure to reply more later...

**Comment by benjamn** [comment]
Please allow me to clarify: I absolutely consider all of you to be actively involved in the beta/RC process. If that wasn't obvious, then I misspoke.

I still want to know what you—as actively involved beta/RC-using developers—want from an official 1.8.1 release (or any other official release).

**Comment by morloy** [comment]
Personally, I’m fine with the current release cycle. I simply feel that there’s a lot of stuff in the pipeline that doesn’t get ready just because MDG isn’t dedicating the resources to Meteor.

For example, Typescript support was announced in the Meteor 1.8 blog post. And it seems to be used internally by MDG, but apparently there’s no time available to clean it up and make it public.

I opened up a PR for source maps in Nov ’18 and didn’t yet receive a single comment from the maintainers. This makes me feel that Meteor simply doesn’t get the attention from MDG that it deserves. Meteor has 40k+ stars and is the best full-stack framework out there, but not even one maintainer working full-time on it (this is my guess and might be wrong, but everybody at MDG seems to be working on Apollo as well). It’s quite sad to see Meteor float in the void without a clear roadmap.

**Comment by yanickrochon** [comment]
I may not have contributed directly to Meteor, but I have been following this project for about 6 years now, and have at least one medium-sized project currently running on some private company server for the past 3 years without failure.

My biggest issue from MeteorJS is how quickly third party projects and libraries fall behind. If I want to start a new project with the current stable release of Meteor, some tutorials no longer work, or the referenced projects are simply dead, not maintained. This is, to me, a project killer. How many times have I heard project managers argue over open source because of "lack of or unreliable long term support"?

Extending the release periods is fine, as it gives enough time for third party projects to keep up to date, but there should be some way to monitor the compatibility. For example, the [accounts-ui](https://github.com/studiointeract/accounts-ui) has not been significantly updated for the past two years, and the [Angular tutorial](https://www.meteor.com/tutorials/angular/creating-an-app) should be called "legacy" because AngularJS is on life support (LTS).

So, release cycles should include :

1. Updating the docs with up-to-date tutorials, using most recent vendor updates (e.g. Angular)
2. Making sure that all the "must have" third party packages are maintained and still actively supported.

These things should get the community more involved, and increase the level of assurance in the Meteor ecosystem.

m2c.

**Comment by brucejo75** [comment]
OK, I am ignorant of your current processes 😺 

I was starting to write my response when I discovered this: https://github.com/meteor/meteor/milestone/61

For me this helps explain why 1.8.1 is not released yet!  That is where I can get specific in my feedback.

## Some tweaks
I also like the model that VSCode uses: https://github.com/Microsoft/vscode/milestones

1. issues in current release (are addressing)
2. issues on deck for future releases (intend to address soon)
3. backlog (would like to fix)

Curating these lists is quite a bit of work.  But they communicate very clearly where future energy is intended to be put into the release.

## Time based releases?
I have observed when projects move to time based releases (e.g. VSCode: monthly, node ~weekly) it has the effect of building confidence with all stakeholders.  I experienced this first hand many times and it is a simple fix that is usually easy to make.

I think Meteor could move to a quarterly release cadence and that would help build confidence in the overall processes.  To make it work you will need to start enforcing some tough standards:
1. the schedule is feature priority 1, nothing can block this.
2. to be in a release, you need to be submitted with zero known bugs.  This makes sure that the release is all about the integration of all features, not the debugging of a feature.
3. if a feature is blocking a release date, it can and will be dropped.
4. there will be internal dates on a milestone (e.g. nothing, absolutely nothing new 30 days prior to release date)

Each quarterly release could drive a theme or 2.  Or a release could simply be about fixing bugs or cleaning up tools.  But each quarter a release is made.

**Comment by StorytellerCZ** [comment]
Would I love to see more communication and resources put towards Meteor? I think that goes without saying for any project I like and follow as intently as Meteor.

Can I personally contribute to the Meteor development? Only very little (busy with my day job and trying to start a business) that I have to blush when I see the `Contributor` tag next to my comments.

I understand the situation that MDG is currently in. My hope is that there is plan to move more resources towards Meteor and Galaxy (would love to see more regions - MongoDB Atlas region editor spoiled me :yum: ) once Apollo reaches some great point. It just means that in the meantime I have to do more (and I would like to if I didn't had more pressing matters, for example I would like to update the MongoDB functionality used by Meteor to allow things like insertMany, etc. in a nice Meteor way).

When it comes to releases I'm only afraid that there is feature creep every release (ie. how 1.7.1 became 1.8, which leads to long times between releases), I'm afraid that we are reaching the same point with current 1.8.1 work. But I'm in the beta process so that isn't that pressing for me. Though I would like to see more PRs merged after 1.8.1 and React packages are have some great [PRs](https://github.com/meteor/react-packages/pulls) ready as well. Still at the end I don't have problem with the speed. At the same time revising the roadmap could be a good practice after each release.

In the end though I think @benjamn gets the short end of the stick here and for my part I'm extremely happy that he is focusing on development because that is where his talents are best used, if that means less communication on general stuff, then that is fine with me.

In the end this topic belongs to the forums (where this was being addressed multiple times) not on GitHub. I sadly don't think that any discussion here is going to get attention of the people that can actually change the current status quo. It is only going to waste @benjamn's time.

**Comment by sakulstra** [comment]
I'm not hopping on the rant train, but can somehow relate to the sentiment of this thread.

It's not sth special I want/need from this release. This release fixed some issues for us(mostly the node upgrade irrc) so we started using one beta in production at some point. More frequent releases would allow us going the beta route more lightweight as now we have to decide between: "upgrade to beta-x containing a bunch of stuff i don't know anything about and have no time to look into" or "miss out on patch releases". I feel like we're in the same situation as we have been in 1.8.

I personally am looking forward to https://github.com/meteor/meteor/blob/devel/Roadmap.md#full-transition-to-npm since the 1.3 blog post. For the last year we couldn't keep up with fixing/forking/inlining dying atmosphere packages we've been relying on. We've been facing similar issues as reported in https://forums.meteor.com/t/testing-meteor-apps-in-2018-a-rant/44662 and ended up not unit testing the meteor-serverside of things at all and using jest + a massive amount of mocking to test the rest. Perhaps that's a completely irrational thought/hope -> I'd just imagine things would be easier if meteor was better integrated in the npm ecosystem.

I don't expect that to happen in the coming releases, but I'd love to see some transparent communication on how to get there / how we as community can help getting there. I guess open prs like this one: #9603 somehow discourage leaps in that direction, although I personally always had a very pleasant and helpful experience with my prs (thank you :clap:).

**Comment by jamesmillerburgess** [comment]
I'm actually quite happy that MDG still supports Meteor as much as they do, even though Apollo has been such a success and investment there must certainly have a much higher ROI. That's not an easy choice to make for a small tech company with limited resources.

As @benjamn suggests, we use the beta versions then they have something we are interested in, so we aren't waiting for an official release before taking advantage of improvements.

The main complaint our developers have about using Meteor is long build times, and thankfully someone from the community has stepped up recently to help alleviate this issue.

**Comment by jamesmillerburgess** [comment]
Actually I should note that @benjamn has been focused on build times and has done amazing work there. I’m just glad that in addition we are benefitting from community involvement on the issue!

**Comment by smeijer** [comment]
@jamesmillerburgess, I very much appreciate @benjamn his work, but unless I've missed something the recent performance improvements are thanks to @zodern and @yorrd.

That being said. @benjamn, In my opinion, the issue lies between the lines of [your post](https://github.com/meteor/meteor/issues/10477#issuecomment-468704213). And thereby it's pretty much impossible for you to fix.

I think I can speak for all of us here when I say I'm thankful for what you are contributing. We realize that you are a busy man and try to do whatever you can.

And that's what our problem is. You're a one-man team. The post you've written is (understandably) very personal. "I've tried, my predictions, I'm sorry...".

I hate it that it's you who's the receiver of this topic. The thing is, we are grateful for your contributions. We are just missing MDG, your colleagues, CEO, marketing, management, communications, other developers/contributors.

Meteor is too big to run by a one-man team.

I fully understand that you can't or won't respond to all posts on the forum and/or github. It's not your job, and it's too time-consuming. And that's why you should have gotten support from other MDG members.

That's the thing. You are awesome! But you can't run this project alone.

So to you personally @benjamn, Thanks for your hard work! I'm grateful. Despite my frustrations towards your employer.

**Comment by jamesmillerburgess** [comment]
@smeijer 

I'm considering Meteor 1.8 with the delayed legacy builds to be recent, but maybe we are thinking in different time frames.

**Comment by morloy** [comment]
@smeijer you got a point here. Meteor is too big of a project for the support it currently gets.
Maybe it’s wrong to blame MDG here. They have venture funding and need to look for themselves how to turn that money into a sustainable business. And maybe that’s not Meteor. (Although in this case, they shouldn’t be running on Galaxy as well.)

I wonder if the support of the community would be large enough to sponsor at least another full-time position, maybe via Patreon? I’d definitely throw in some resources to make Meteor a thriving project!

**Comment by KoenLav** [comment]
@benjamn just to make sure; this is definitely not criticism on your work; I'm addressing MDG as a whole.

I guess I did misinterpret your comment, apologies for that. 

I'm going to get back to everyone's questions and suggestions within a week; it's good to see there is significant interest from people to improve upon the current situation. 

I don't think things are bad, just that they could get better and some structure/planning/communication could help with that.

With regard to 1.8.1; we've been using it in production for a couple of weeks without any issues (I also tested Cordova, but not yet the HCP changes), but shall we move the discussion specific to 1.8.1 to the release thread?

**Comment by yorrd** [comment]
Honestly guys, not trying to make this thread longer than it needs to be. This is going to be a +1 with a somewhat different perspective. I love Meteor, it cuts down on initial development time like no other tool to date. And believe me, I have done a lot of research.

@benjamn (sorry for spam tagging) there are two questions that I'm currently asking myself that both add up to the same issue: I just don't know what to expect from the future out of MDG right now. If I would and the goals would be useful to us (they have been so far), we as Adornis would invest heavily in community maintained packages and pull requests towards Meteor AND Apollo. We like and support your move to build a new data system (sorry for the stupid expression) and would like to be a part of it because you and the rest of MDG have proven to be a reliable source of innovation and thought-through software.

1) what is the future of Apollo + Meteor, does MDG have a focus on bringing reactive data and optimistic UI (minimongo like) to Apollo? We're trying to hack together something that works inside Meteor and we're going to release it open sourced when we have something semi-stable. I'm skeptical towards this because live queries and optimistic UI (while certainly being possible) don't seem to be a focus right now if I interpret you guys on slack and github correctly
2) what is the future of an officially supported TypeScript package? I started working on rewriting the whole barbatus package in order to support the new incremental compiling stuff from 3.3, but I'm just not sure if our work is going to be done in vain because MDG release a better version a few weeks later anyways

Conclusive answers to these two questions would help us plan more long-term and allocate resources towards open source software, mainly Meteor, Apollo and lit-html. 

And to clarify: there is NOTHING WRONG with your release process and the way you do open-source, especially you personally, benjamn (like others have stated). Meteor for me is a prime example how a community engages and comes together around development of the platform. The ONLY thing we as a company are desperately in need of atm is clear communication on the current priorities. I'm not even sure who within MDG would make that overarching decision to be honest, and I think that's part of the problem. 

Thanks for your time, I know this is a shitty subject for you guys and there are no easy answers here.

**Comment by yorrd** [comment]
@morloy we're interested in participating in funding or helping with development on Meteor, assuming that is what MDG would sign off on and support as well

**Comment by brucejo75** [comment]
I looked at the [issues that are holding up 1.8.1](https://github.com/meteor/meteor/milestone/61) and they are a little surprising for this release:
1. They are pretty stale, a lot of them have not been commented on this year.
2. There are requests for PRs, but the community has not stepped up.
3. Some of them seem like internal / annoying errors that are kind of edge cases that should not delay the release.
4. Some of them are waiting on MDG review input.

Are these issues really holding up 1.8.1?  It seems like they are not?

### Candidates for punting (not high priority and need work)
#10427 [WIP] Allow interrupting delayed builds.
#10381 Changing port in development can cause `MongoError`
#10366 Lazy compilation errors swallowed in certain circumstances
#10345 [wip] Remove multiple restart feature
#10230 Meteor including unnecessary node_modules
#10341 Source maps in standard-minifier-js
#10237 Multiple concurrent deploys to Galaxy puts deploy steps in a loop

### Issues that need some action
#10277 Hot Code Push broken on iOS [is this resolved?]

**Comment by sebakerckhof** [comment]
Thanks @benjamn for taking some time to get involved in this discussion, however:

> I've tried in the past to make predictions about when official releases will happen, and it just doesn't work. 

I don't think this is the issue at hand here. It's not so much about missed ETA's. We're all developers here so everyone understands how relative time predictions can be (when the 1.8.1 thread started my bet was actually on it being renamed to 1.9 around March anyway ;) ). But the complaints are about a total lack of communication.

MDG has always been untransparent about their future plans and roadmaps. Understandable, publishing a roadmap has people thinking you commit to these things which you might not be able to achieve when plans inevitably change. This wasn't so much an issue when we knew that the main focus was on Meteor. But since 99% of the company is focused on Apollo this brings a lot of uncertainty. I think this uncertainty is the core of the problem and can be alleviated with even a little bit of communication. There's people building their products on Meteor without a clue if MDG is planning to keep Meteor alive. That can feel frightening. Seeing there's been very little activity since December on the Meteor repository and you working on Apollo doesn't help.

Now we can say that MDG doesn't owe anyone anything and it's open source so everyone can contribute if they don't like it. But MDG is also not very welcoming towards contributors. There's PR's that have been open for a year without any comment of any MDG member. This is of course partially due to an understaffed team maintaining Meteor, but even a small comment stating your thoughts would be so much more motivating for contributors and will in the long run benefit everyone.

**Comment by amerzad** [comment]
I know my words may just be wrong, completely out of context or plain stupid or theoretical, but here is what I have in mind.

I personally have developed several apps using Meteor, and some of them are in production, and the operations guy love Meteor because it is the best scaleable solution we have compared to several other frameworks, and it actually allowed for a very big productivity boost for simple solutions done by very newbie developers with training of less than 4 months, and some of these solutions are in production for more than 4 years without any serious problems.

Yet, by time, Meteor apps was always in danger of being replaced with something else because some developers didn't like that so much magic in Meteor, but that was not the actual problem with them.

The actual problems is that, they think having Meteor in their CVs is not that impressive like having Django, Laravel, or even Express.js because these are the ones they are hearing about everywhere, and they are the ones which big companies are doing their projects with, and because Meteor was very easy and powerful to the degree that they started thinking that it cannot differentiate strong from weak developers!

Yes, they are guilty of following their benefits over the benefit of their projects and companies they are working for, and even the management is guilty of preferring the developer experience over the actual benefit and economics, but that is the reality, at least according to my experience.

In the Islamic literature, there is a very known Sufi scholar/philosopher called: Ibn Ataa Allah Assakandary, who wrote a book of wisdom, which I had read it from the age of 15, and I like one wisdom very much and feel it is very very real, which says: Oh son of Adam, nothing have driven you like delusion.

So, how to make the hype active again for Meteor so new/old developers will be willing to spend their time and learn Meteor and participate in the forums or at least increase the download rate and make them press than star button?

I feel it is just that problem with all the guys who complained about the silence.

But for my self I would say, silence is good. I personally look at things from the perspective that I will die tomorrow. Nothing is more important than being quiet as much as possible so we can listen to the reality. And the reality I see is that, Meteor is still relevant to my work as it was from five years with no future guarantees, but I have one new catch, big changes are up ahead, mostly driven by AI, WebAssembly and SaaS, and these things are the only things which worries me about Meteor future, and my career in whole as a Meteor developer.

But for others, I think they need some noise, and everything they want is some noise, so they can tell others and themselves that Meteor is alive. I personally think Meteor is still the best solution for prototype speed production development which can take you far enough before you will need to hit yourself with huge development costs of optimizing the performance. And that in my opinion is good enough for this age where 91% of startups are failing.

I am not saying there is nothing missing in Meteor or I am not annoyed at all by the very long standing feature requests, but I am saying, we (Meteor community and MDG) need to make that noise and do something which can impress people every while, or they will think we are dead, and their delusion could cause us to actually die.

Maybe I am the last one to say that due to how much little I contributed to the community, yet, be sure I fought many battles which Meteor was the flag I was holding, yet, pressure in the Middle East area is just huge, and humanity is writing a bad page of history in my region, so I don't expect this to change much, sorry, I wish I could do more, but at least I am doing some noise :wink: 

I am sorry if I offended anybody, I am almost completely stranger to your culture, and I hope I didn't do harm more than benefit.

**Comment by rj-david** [comment]
Summarizing the discussions, there are 3 major issues in this thread:

1. Communication

2. Resources for development

3. Confidence on Meteor's future (in effect, can be a product of number 1 and 2 above)

**Comment by aliogaili** [comment]
Good summary.  

Actually 50% of the forum posts (and please keep those post at the forum where they belong and let the devs do their work), are all the same:

1. Is X dead 
2. Will X be alive Y years from now 
3. Why my feature (that no one else cares about) not being addressed 
4. If the dev go silent for a week, then panic mode is on. It's the twitter age and nobody goes silent..
5. X was really good (just for me) Y years ago, why can't we go back?
6. and devs favourite question, I want a unix time stamp of the next release date

Despite the multiple times MDG confirmed verbally (and in code) their commitment to Meteor long-term, despite the many successful companies who are also committed to support meteor long-term, and despite the active repo with the interesting recent merges pertaining to build speed, people still need more social proof. 

So in order to resolve this, I suggest public ceremony where MDG team does a wedding like voes that they'll be committed to commit free code to the repo for the rest of their existence. 

Ok more seriously, I think adding more maintainers to the repo would help. But I really think this post should have been in the forum.

**Comment by KoenLav** [comment]
@aliogaili those posts turn into "is Meteor dead" exactly because they are on the forums (I think). Also most of them don't raise any concrete suggestions for improving things (aside from asking more commitment in word or actions from MDG). 

This issue is intended to improve communication (and/or planning/structure) to get a more fruitful collaboration with the community as a whole.

Several people who have been actively involved in the development of Meteor have already chimed in, with either helpful points of view or even very concrete suggestions to (further) improve the collaboration.

A forum, by its definition, is a place for discussion, whereas an issue tracker is aimed at pinpointing and resolving issues.

That's the reason I posted this over here. I can see that this is debatable, but let's not turn this into a discussion (whether it's a discussion about what is the actual issue here or where it should be posted); let's focus on identifying possible improvements.

ps should MDG think this is better suited in the forums then I'm sure all of us won't mind :)

**Comment by aliogaili** [comment]
I see your point and there might be truth in it. I just hope this thread stays on-point, positive and constructive.

**Comment by coagmano** [comment]
Big +1 here for an update schedule. Even if it's just to create in impetus to check and merge PRs.

As much as I don't understand the internet's desire for constant commits to a project, as long as there's updates coming through on a schedule we can point to that as a strong signal that things are okay.

**Comment by MastroLindus** [comment]
Just adding my 2 cents, mostly to agree with the others:
I am very grateful to @benjamn for his hard work. I truly am.

However as the person that needs to take decisions on which technology to rely on for my business, I cannot deny to be concerned about the current situation.
A huge framework like Meteor with only one main active developer maintaining it, it's a concern. Especially if in the last months that developer had a bunch of commits for meteor, and 40-50 times that for Apollo.

The problem is not what meteor doesn't support now. I think it is quite in good shape, (even if certain things would really be appreciated, I for one am also really waiting for official typescript support, or better support for the last mongo features without having to rely constantly on rawCollection()).
The problem is how much trust you can put in a technology that doesn't move forward very fast, in a world that is running at a ludicrous speed.

When as a decision-maker I need to pick one technology, I try not to pick the one with the best features. I try to pick the one that is more alive, thriving, in terms of community, maintainers, clear vision for the future, etc.

I really loved Meteor so far, but that love is mostly coming from how it used to be 4-5 years ago  than now.

So what would I like to see from Meteor to help the situation?
1)Either invest more resources (especially active maintainers) in the project, or help the most active members here to step in if it's not possible
2)Release some form of regular communication (blog, wiki roadmap, github issues, whatever) maybe similar to what typescript and visual code's projects are doing, that states what is your vision, where do you want to go, what do you want for meteor at the end of the year, or so. It can also be "we just want to keep it in the current state, and maybe release some critical fixes if necessary". But at least it's clear and people can plan and decide accordingly what to do.

Little (maybe unrelated) disclaimer about Apollo:
I am very happy for MDG to be excited and investing a lot in Apollo, and I wish it to be very successful for you guys.
However I really don't think it fits many projects, especially small-medium projects, where the simplicity of Meteor is really the selling point. So I am not particularly excited about possible apollo-meteor merges, integrations and such, as I think they fit really different kind of projects. GraphQL is still far from being the industry standard and I would like to not have buy-in into it in order to enjoy all the other amazing tools you have built here.
I am just saying this to state that I really would like to see a future for Meteor as it's on thing, and not having to end up as a package for some other frameworks because of the lack of support.



P.S.: I still want to thank @benjamn , from developer to developer I appreciate the enormous effort you have put in this project for everybody else benefit. Being a one-man-team  I just wish you didn't have to do everything by yourself, it's not good for you, it's not good for the community, it's not good for Meteor.
And thank also to all the others (@yorrd @hwillson and everybody else) for the contributions, you guys are awesome.

**Comment by aliogaili** [comment]
Very well said @MastroLindus!

I also share your perspective on Apollo, while I truly admire the engineering work on Apollo, I just don't see it reaching the level of impact that is has been hyped to be (REST replacement). In fact, I don't even think it'll be as impactful as Meteor once the hype settle. 

Meteor is at the core of many successful businesses and open source repos. We're using to power 5 enterprise apps and counting, we're using [RocketChat](https://rocket.chat/) and  [Wekan](https://github.com/wekan/wekan) for task management which also powered by Meteor! I personally know that RocketChat is being deployed in [Canadian](https://events.linuxfoundation.org/wp-content/uploads/2017/11/BoF-Using-OSS-to-Build-the-Digital-Workspace-for-the-Government-of-Canada-S%C3%A9bastien-Lemay-Treasury-Board-of-Canada-Secretariat.pdf) and UAE government offices since it can be hosted internally. 

Furthermore, many successful startups such [Quali](https://www.qualia.com/) and [Classcraft](https://www.classcraft.com/) are very clear about the role Meteor played in their success. In addition there is [VulcanJS](http://vulcanjs.org/) that is power another layer of the web apps and sites such as [LessWrong](https://www.lesswrong.com/).

I think a 40K star framework with such an impact does deserve more attention.

**Comment by KoenLav** [comment]
@benjamn 

> What specifically matters to you about the final official Meteor 1.8.1 release that isn't satisfied by the beta releases?
> 
> From my perspective, if you're a Meteor developer in 2019 and you don't know about the beta/RC release cycle, you're probably not actively invested in Meteor development, or in Meteor's future. And that's fine, but it means you're enjoying the benefits of free software without helping to ensure its longevity. If the relative infrequency of official releases bothers you, you've probably already figured out that the beta/RC cycle is the place to be.

Using beta versions in production should not be necessary. The Node fixes (Fiber bomb) and Cordova changes (incl. HCP fixes) are the main reason for us to want to move ahead. We've definitely 'figured this out', but that isn't to say we're happy with it.

> The long release cycle exists so that folks who need the new stuff can begin using it sooner, and those who can wait for the official release will (eventually) benefit from all the testing and validation that went into the beta/RC process.

This is a nice way of framing it after the fact, but the release post mentions a 'quick follow up'. So either there was a different planning and the community was misinformed, or there was a planning but it got thrown out of the window, or there really wasn't (isn't) any planning at all (my guess). In any of these cases more communication would have helped.

Just to make sure; I'm not blaming anyone, I'm want to isolate the 'issue' and work together in resolving it.

> I've tried in the past to make predictions about when official releases will happen, and it just doesn't work. Unexpected contingencies like Cordova bugs emerge at surprising times, and demand my/our attention. I'm sorry that I've ignored requests for new predictions, but please understand: my predictions will always be pretty vague, because I don't fully understand what you're (collectively) looking for in an official release.

It would be nice to get an ETA on every release, but everyone understands this cannot be accurately predicted. What we do however know for sure that if there is no (communication about) planning at all then this produces uncertainty.

Even when asked about some communication about planning (whether expressed in time or work to be done) we (and I specifically) got zero reply:

> @benjamn any rough expected release date for 1.8.1?
> 
> Or is there a concrete list of points which need to be addressed before release?

Maybe I should not have addressed you specifically (but then who?), but this would have definitely been a nice time to turn the question around and ask me/us/the community whether we thought it was ready for release.

Again, I'm really not trying to point any fingers, it's just about establishing where we can improve the proces (and maybe also even about finding out who is/should be responsible).

@brucejo75 created a nice overview on which outstanding 1.8.1 issues he perceives to be non-blocking, but let’s move this discussion back to the Meteor 1.8.1 release thread ;)

> If you have a specific feature or bug fix that you'd like to see released officially, and you think it's ready to ship, I would really appreciate that kind of feedback. Obviously those requests need to be balanced against any bugs that other folks are expecting to see fixed, but concrete reasons to cut an official release would help a lot. Right now the 1.8.1 release is in a sort of limbo where I'm waiting to be sure the update process will be smooth, and passive developers will be glad they updated. Are we there yet? I genuinely need your help to make that call.

After asking this question I think the community, within hours, provided us with sufficient feedback to make a decision on this. Apparently merely asking the question sufficed. If there was some kind of process in place that would make this kind of collaboration more explicit I think this would help you as a person, MDG as a company and the community as a whole.

I hope we can agree the way in which MDG collaborates with the community can be improved; please let me know whether this is the case. If so we can quickly turn this thing around; instead of looking back at the past we should be working towards concrete improvements in the future.

@benjamn some people also raised the concern that this topic might just distract you from your other responsibilities and/or even that you're not the person we (as a community) should be addressing. If this is the case, please let me know and I'd happily discuss this with anyone from the MDG team.

I'll address a few more replies specifically before I attempt to provide a list of concrete action points which we might want to (eventually) split up into separate issues/threads/working groups.

**Comment by KoenLav** [comment]
@StorytellerCZ it’s definitely not my intention to waste @benjamn’s time, but I do understand your point in that (if we look solely at MDG) then we shouldn’t (just) be talking to @benjamn (which is why I addressed MDG as a whole).

However I don’t think we should look at MDG alone; I think there’s a lot the community can do. Hence my desired end-result for this issue would be twofold:
A concrete list of points which allow the community to help MDG in safeguarding the future of Meteor;
A concrete list of points which allow MDG to help the community in safeguarding the future of Meteor.

My assumption is that MDG will not be dedicating more resources to the development of Meteor in the foreseeable future, so we should make the best with whatever resources we currently have.

**Comment by KoenLav** [comment]
@smeijer thanks for this exhaustive overview of the history of Meteor.

My main takeaway from your numbers if that, while MDG has promised to (keep on) committing significant resources to Meteor, this does not appear to have happened.

Personally I wouldn’t say that I’ve lost faith already. While I definitely think that creating a fork is an option, I would see that as a last resort after we exhaust all other options; forking such a large open source project will require significant investments from any future maintainers.

**Comment by KoenLav** [comment]
@yanickrochon while I agree that there are a lot of Meteor packages which are outdated, most of them still provide us with what we need. Packages can be easily forked and should be maintained by the community. In addition to that I think the community has been slowly moving from Meteor packages to npm packages.

I do think it could be a great idea to create a ‘Meteor Community’ GitHub organization (just like React Native Community) which could become responsible for Meteor packages which cannot be (easily) replaced with an npm variant, but if people are interested in that I think we should create another issue and discuss that over there, as over here I’d like to focus on how MDG and the community can work together to strengthen the core of Meteor (not the package ecosystem surrounding the core).

Apologies if this was not clear to begin with.

**Comment by KoenLav** [comment]
After reading all replies and structuring them I came up with the following issue description:

**The Issue**
The **resources** which MDG currently **commits** to Meteor, a rather **large** open-source framework, combined with the (lack of) **communication** around this, causes **uncertainty** which leaves the **community** with insufficient **confidence** in the future of Meteor.

**Key terms**
- Resources: Meteor is too big to run by a one-man team;
- Communication: Lack of communication causes uncertainty;
- Planning: Meteor’s uncertain future leads to insufficient confidence.

Next I'll try to summarize the suggestions for improvements in a clear and concise way, outlining pros and cons.

**Comment by KoenLav** [comment]
### Communication-based improvements

**A. Regular release cycle**
Raised a couple of times; I think we can all agree that having a regular release cycle will make it very clear to everyone when a release is to be expected. If there is a specific feature/fix the community wants to see in a release, they are incentivized to make sure it’s ready to ship ahead of the release date. When @benjamn (or anyone else from MDG) is unsure on whether the release is ‘there yet’ they have an incentive to ask the community sooner rather than later.

I would say the community should be able to expect at least one release every 3 months (whether that’s a major, minor or patch release doesn’t matter, in my opinion). The community also shouldn’t care whether it’s 2 months or 4 months; it should just be a guiding principle (there should be a release somewhere in Q1, somewhere in Q2) and there are no promises in terms of what’s included in the release.

Concrete: are there any reasons not to do this?

**B. Process: clarify roles & responsibilities**
Who is responsible for the release process? Is MDG waiting on the community to sign off on a release? Or is the community waiting on MDG? Should a Project Manager be assigned? Can @benjamn take up all of this work, or should he receive help from MDG/the community?

Concrete: MDG should clarify this.

**C. Hype: make noise!**
@amerzad mentions Meteor isn’t “hyped” enough (anymore), which doesn’t help the confidence in the future of Meteor. Although this is not my major concern I do agree that the relative silence around Meteor isn’t helping things. This kind of effort has been discussed a couple of times on the forums already and every time the conclusion was that the community should take the lead here. 

Concrete: the community should pick this up.

Who wants to take the lead and create a specific issue/thread/working group for this?

### Resource-based improvements

**D. Community: welcoming maintainers**
@sebakerckhof puts it nicely in my opinion; “MDG has not been very welcoming towards contributors”. Many PRs have been left unanswered and end up being stale. Maybe we should assign community maintainers who want to take responsibility for this?

In recent news @benjamn has just asked @zodern (responsibly for the majority of build time improvements over the past months) to join as a maintainer (I think?), so this is already a nice step in the right direction.

Concrete: MDG should pick up the ball here.

**E. Project size: reducing the surface area**
Some time ago React Native decided to remove quite a few Components from their core in order to reduce the surface area and get a better grasp on the project. I reckon the package architecture of Meteor would play with something like this quite nicely.

We could create a plan for slimming the Meteor core and transferring (significant) parts of the core to a community-maintained repository. Possible candidates for removal from core:
- Cordova integration;
- Compilers (e.g. TypeScript);
- …?

Concrete: I’m not sure whether this is the right direction, so I’d like to get some feedback on this approach.

**F. Crowd: funding or sourcing**
@yorrd mentioned that they are interested in funding and/or helping in development for Meteor. I guess we could call this crowdfunding/crowdsourcing. I definitely think this is an interesting idea, but I do agree it would be preferable if MDG signed off on this.

Concrete: MDG should let us know whether they are interested in this

**G. MDG: increases their commitment**
At this point I agree with @smeijer; I don’t expect this to happen any time soon, but boy would it be nice. It would be nice however to know, for the foreseeable future, how much capacity they are willing/planning/hoping to invest into Meteor.

In my opinion Project Management, at this point, is as important as Development.

Concrete: we would appreciate some insights into the capacity planning from MDG.

**H. Forking: community-maintained fork**
@smeijer suggests forking Meteor as he has lost faith in MDG. Personally I haven’t lost faith, I just think that there’s a lot lacking in the Project Management department. If we can somehow pick up the ball with regard to this then I think we should be fine.

I think the pros and cons for forking are quite similar; while we could shape the future of Meteor ourselves, we would have to shape the future of Meteor ourselves. While we could pick up some speed in development, we would have to pick up speed in development (otherwise it would die a slow death).

Who will be taking the lead? Who will commit to doing what? I think the Meteor community, at this point, is not properly organized to take over the development of this project.

**I. GitHub: Meteor Community Organization**
When discussing these improvements I notice that many of these points require us, as a community, to be much more organized than we currently are.

Concrete: we should set up a Meteor Community organization.

### TODO
To be honest, I'm not sure what should be the way forward from this point, but I do know that:
- I'd like feedback on the concrete points mentioned above from both the community and MDG;
- At some point we should definitely split this issue into smaller units of work.

Thank all of you for contributing so far and I'm happy to hear your thoughts!

ps apologies if I misrepresented your views and/or missed out on some crucial points, I tried to make this as concise as possible.

**Comment by rodrigok** [comment]
Thanks for mentioning Rocket.Chat here @aliogaili.

We, at Rocket.Chat, decided to move away from meteor as a mid term project duo to all reasons already mentioned here and more:
1. It's hard to improve performance using meteor, we are already moving away from DDP and trying to stop using fibers.
2. We have an issue related to mongodb driver, solved already but not released since november https://github.com/meteor/meteor/issues/10362
3. Microservices is our future and Meteor isn't prepared for it
4. Framwework development is very slow and I'm afraid it will keep not solving simple problemas fast enough 

We already started this movement of moving our code structure from packages to imports and removing the global variables to be as close as possible to a pure NodeJS project, the next steps are stop using meteor's collections in favor of raw collections and refactor the code to stop using any specific meteor code.

As an Open Source project we understand how hard is to maintain a project without a revenue source, I thought that Galaxy would solve this problem for Meteor, but it seems to not. Some times it's better to try to find a way to charge for the service and keep the good pace of the project rather than slow down. I know it's a challenge!

I hope things will improve here and we may not move away from Meteor at all, but now it's our plan for this year.

**Comment by aliogaili** [comment]
@rodrigok I hear you, thanks for sharing, it'll be a big loss for Meteor.

That would be a considerable refactoring given the size of your code base and your dependency on blaze, are you guys tracking this migration somewhere on GitHub? I really hope your experience improves with Meteor (specially 2 & 4) and you keep growing with it.

RocketChat is awesome, keep it up guys!

ps the folks working in the government in UAE really loved rocket chat but they said they're early waiting fo threads, so I thought I'd pass it to you.

**Comment by rodrigok** [comment]
@aliogaili we do not have a single issue to track this, it was splitted in several projects in our internal asana account and things keep moving, we start the issues and PRs as soon we decide to work in one of the steps or even decide what is the next step.

The initial item was the [removal of globals](https://github.com/RocketChat/Rocket.Chat/issues/12369) with 223 PRs merged and thousand of files changed.

The second one is a [pull request](https://github.com/RocketChat/Rocket.Chat/pull/13525) containing the result of a script we created to move everything from packages to imports and resolve the imports. This PR is waiting the merge of 2 big things for 1.0, [Federation](https://github.com/RocketChat/Rocket.Chat/pull/12370) and [Threads](https://github.com/RocketChat/Rocket.Chat/pull/13541) to be able to run the script again and convert these 2 additions as well.

The next step will be a refactor in our models using TypeScript and a microservice framework. But we need to accommodate it in our roadmap before have public tracking.

About threads, we started with a [contribution](https://github.com/RocketChat/Rocket.Chat/pull/11803) from the community and created an internal branch to add some improvements. This will be our first implementation for threads, it will be shipped within version 1.0.

**Comment by benjamn** [comment]
@rodrigok I'm going to hide both of our comments as off-topic, because they have little to do with planning or communication. However, I think your feedback deserves a response, even if it's not appropriate for this thread.

Migrating away from Meteor is a uniquely effective way to remind yourself and your team exactly what benefits Meteor provides. You will find yourselves responsible for many more aspects of your project's design and configuration than before, but you won't have anyone else to ask for help (or to blame). If that feels liberating, great! You made the right decision. If your development velocity slows to a crawl because there's so much more to worry about, you're always welcome back. ✌️

1. Performance doesn't get any easier for non-Meteor real-time applications, I'm afraid.
2. @sebakerckhof gave you [direct advice](https://github.com/meteor/meteor/issues/10362#issuecomment-447414323) about how to fix the problem in the meantime, and you brushed it off.
3. If you just mean Meteor doesn't do anything to make microservices architecture easier, that's a fair criticism, but I don't think abandoning Meteor will help, either.
4. You're now going to be responsible for solving all the simple (and hard) problems, so I hope you're right about being able to do it quickly enough.

**Comment by KoenLav** [comment]
@benjamn thanks for keeping things on point.

Just to make things extra clear for everyone; this issue is intended to facilitate a constructive dialogue between MDG and the community with the goal of arriving at a better collaborative process between the two.

If you're leaving Meteor or if you think 'Meteor is dead' and as a consequence are no longer willing to invest time in improving Meteor then this is not the issue for you.

I honestly appreciate all insights, but within this issue things should be aimed at achieving a better collaborative process; anything else is out of scope.

**Comment by rodrigok** [comment]
Sorry for the "branch" I did open here, just want to clarify the position of Rocket.Chat since it was mentioned here.

Related to how to handle the communication with the community, one of the things that did help us was to keep a fixed release cycle, meaning we have a new release every month and the clients/community knows the time to expect to have the merged things released as GitLab does as well.

This fixed release cycle helps the community track what we have planned for the next releases, since we can't add too much items or we would need to pass the release date.

But, for sure, the most important here is know what MDG wants for this project, either open for more community maintainers and pass the roadmap decisions to the maintainers or make clear if they want to invest more (with developers) in this project.

**Comment by coagmano** [comment]
I know there's been some discussion of setting up a Meteor community in the past ([forum example](https://forums.meteor.com/t/time-for-a-meteor-collective-for-community-packages/38230)

**Comment by coagmano** [comment]
I know there's been some discussion of setting up a Meteor community in the past ([forum example](https://forums.meteor.com/t/time-for-a-meteor-collective-for-community-packages/38230))

While I don't want to own it, I would happily support an effort to make this happen. Happy to help slowly fork, review PRs and issues, republish to atmosphere etc anytime I get a spare moment

**Comment by amerzad** [comment]
I welcome all of the suggestions of @KoenLav https://github.com/meteor/meteor/issues/10477#issuecomment-469218868_

I personally think MDG is doing what they think is right for Meteor because they are the most knowledgeable people about Meteor and how importance stability is, and I have no reason to doubt their interest in Meteor, but what is happening is just not enough for the rest of the community, especially the hype thing.

I personally prefer having a fork with all the instability it can bring with the so many PRs to be merged, just to let everybody have more options, and let them realize how hard it is to advance such a framework, and at the same time, let them see what they will need to do on their own when they leave MDG, who knows, maybe the community will be powerful enough to do just fine, and I have hope in that.

Maybe it should be named: meteor-next, so meteor will be the stable one, and meteor-next will be the edgy one. I just hope MDG do that fork and lead this step, as this will make it much more easier and noisy.


--------------------------------------------------------------------------------------
--------------------------------------------------------------------------------------
What follows is off topic or should be TLDR;ed or even deleted by the moderator if it is not serving any good purpose, and please, don't read it if you don't have free time, and don't reply it on this thread. I personally think it is deeply related to the topic, but I think not everybody will share this with me.


I am still believing in the benefit of Meteor especially for small and medium sized startups, the problem is that, those startups are very affected by hype, and if we don't have any more energy to make enough noise, maybe those startups will never correctly understand their options, and they will repeat the same mistakes again and again. I just hope someone wise enough (like @benjamn) to handle it and prevent the noise from leaking to the miraculous quality/stability of Meteor. I feel Meteor is light ahead of its time when it started, and it is still relevant without any serious backward compatibility problems, and that is miraculous to happen in the so violent ocean of JavaScript Fatigue exited in the last few years.

What Rocket.Chat did is exactly the kind of delusion which I was talking about, especially with those 4 weak points mentioned. I do agree totally with the answers of @benjamn, and I would say that the right action was to invest in more Meteor packages or npm packages to support the framework or even help starting a new fork of Meteor - the framework which is yet to have any real efficient alternative other than doing almost everything with yourself without thinking about the future for others.

Rocket.Chat decision is not mine, and maybe it is completely right, maybe I am the deluded one, but for me it is unlikely to be right especially because of how much weak these points are, and such weak reasoning is many times is the actual cause of failure, and from this alone I think the community should know how hype is important to guide people to their own goodness.

I previously suggested Meteor for a project, but the developer refused to use it, and he used Express.js because he think Express.js will be more stable and performant to the project on the long term, so he did a lot of things on his own, and instead of finishing his work in 4 months, he finished it in more than a year, and at the end, they discovered that, the whole idea of the business may not work at all, and after spending millions they couldn't have more than 500 active users!!! What a waste of stability and performance :smile: , but I think the actual problem is that the developer was not thinking for the benefit of the project as much thinking about the benefits of his CV and his "peace of mind", yet, he forgot that, a successful project is much much better for his CV than a framework and number of code lines, and if he finished his work in 4 months, maybe they will be able to redirect some of the invested money to do some business changes to survive.

Yesterday, another developer complained about how long Webpack is taking to compile his almost finished TypeScript CRA project after each change no matter how small it is, yet, he forgot that he criticized Meteor of being slow during development! It is simple, when your code has a lot to offer, it will be slow, and Meteor has a lot to offer from the beginning, and if you don't need what it offers, then do not use it.

Yet, I think forking does not mean losing faith in MDG, it means the community is good enough to handle things on its own, and I believe if the community proved that, MDG itself will not have a problem to join the efforts, and that what we expect from other companies who are either using or benefiting from Meteor or are interested in it for the future.

Meteor is not dead (and will never be as long as its community is believing in its unique benefits), but maybe the community is sick due to the so huge delusional mentality people are having. I think humanity should save itself. And we need to do something at least to save our investment in Meteor, and protect new developers from reinventing the wheel (like monkeys? maybe Evolution is just a delusion after all).

**Comment by aliogaili** [comment]
@rodrigok thanks again for the info. 

I agree with what @amerzad is saying though, the users I talked  never mentioned to the issues you listed, they didn't know/care what tech it was build on, they are just looking for more features. 

Meteor got you where you're today with limited budget due to its speed of development and cost savings it provides. I genuinely believe if this communication "bug" is improved and the community can get more clarity going forward, Meteor can get you even further, faster and cheaper.

**Comment by KoenLav** [comment]
Guys, I think both @benjamn and I made it clear this isn't the place for a discussion on whether you should, as a startup or any other company, pick/invest in Meteor.

I'm happy to hear you believe in Meteor, but replies like this will only trigger others to say the opposite and get a discussion started, for which this issue is not the right place. 

This issue already has a huge scope as-is, so please try to make sure every bit of your comment is advancing the resolution of this issue. 

I know all the other stuff is also interesting to talk about, but it will only clutter the discussion.

Please :)

**Comment by aliogaili** [comment]
Alright I agree and I was the first to point out we need to stay on point, but that discussion in my opinion also highlight why this communication issue needs to be improved, after all you also started the thread for similar feelings/thoughts I assume.

Anyway I'll stop here and let others chime in more productively.

**Comment by KoenLav** [comment]
@aliogaili definitely. There's nothing wrong specifically with your reply, my reply was more aimed at preventing others from continuing on the same subject (Meteor yes/no).

**Comment by moberegger** [comment]
RE: Community engagement

Perhaps as a low hanging fruit, we can introduce a label that we can placed on issues or feature requests to indicate tasks that would good for new contributors to work on (if this already exists, ignore me!) Meteor is a large project and has a big barrier to entry that makes it too overwhelming for many to get started. If we need more contributors, we should do what we can to make that barrier easier to get through.

Keep in mind I am only speaking from experience, so these are just my observations. Personally speaking, some of the issues with the build system or internals go way over my head and I have a huge feeling of helplessness to assist in moving things forward.

A long time ago, I submitted a PR to add some more policies to the BrowserPolicy package. When I first started, I was completely freaked out! To be honest, if I didn't need it there for my job at the time, I probably wouldn't have been able to break through that initial fear. It did end up being pretty straightforward, proving that my initial perception was wrong, but had I not had an impetus to make the change, I genuinely think I would acquiesced immediately.

I wonder if many others feel the same way, even if they are more than capable of contributing.

**Comment by SimonSimCity** [comment]
A fair argument which nobody has mentioned here before:

MDG has been listening to the feature requests the community has had and has, due to the fact that there are so many of them, forked them out into a separate repository: https://github.com/meteor/meteor-feature-requests If you're looking for a specific feature, chances are high it's already requested and got some votes there - maybe ways for implementation are also discussed.

MDG started flagging some issues as `pull-requests-encouraged`. In my eyes, they thereby tell us (as the Meteor community) to start working on it. These are things they see fitting into their future plan of Meteor. See the list of issues here: https://github.com/meteor/meteor-feature-requests/issues?q=is%3Aopen+is%3Aissue+label%3Apull-requests-encouraged

I, too, am guilty of not working on these, but in times where everyone is complaining about MDG being silent, I felt compelled to put this up as an argument for MDG - even though I also see room for improvement.

**Comment by mrmowgli** [comment]
I'm going to add my two cents here, mostly because I still believe that Meteor is a great framework.

I think that @KoenLav is definitely shaping a useful conversation here. I'm personally on board and glad the conversation is happening and clearly productive.  I understand the topic was changed a bit but there are a few things I'd like to mention as a long time user of Meteor. 

I have seen a lot of change in MDG after the creation of Galaxy. I would say there was a shift away from community driven development. Definitely a lot of people left as part of the Meteor hosted community projects shutdown.  I've also been watching the changes related to Apollo, and wondering where MDG was going with it.  Things just generally seem to be losing focus.

One of the core things I loved about the initial incarnation of Meteor was the fact that it was clearly ahead of the game and **opinionated**. Learn the framework, become productive. Magic! Over time though things inside of Meteors original opinionated framework started getting dropped.  Blaze is a great example.  Blaze is a great system and deserves a lot of respect. However Blaze has been limping along with no new development for years and hasn't been released as an independent project that people could actually adopt and improve. What's the replacement?  Angular, React, Vue? Whatever you like, including Blaze.  In my opinion, constantly trying to support all of the possible combinations also means it's hard to rally around best practices, or solving common problems for the entire community.

So, I have a few suggestions in addition to the ones suggested above:
- Become opinionated again: Figure out as a community what the best path to making an application that can grow if it gets traction and own it.  Is it React, Coffeescript and Apollo?  Or is it Angular, Typescript and MongoDB?  Whatever the solution is, focus on that narrative and trim down the things a new adopter has to learn to be productive.
- Standardize on NPM: Is Atmosphere a reliable source? Atmosphere isn't owned by MDG, and the owners aren't active.  Become opinionated, force NPM as a breaking change. Define a standard way of finding and identifying Meteor specific packages in NPM and push Atmosphere package owners to move.
- Use Fedora and  Red Hat Enterprise as an example: Fedora is the partially supported child that creates the innovation and community can easily contribute to and maintain.  Red Hat Enterprise cherry picks and hardens the useful innovations from Fedora.  Meteor Community could literally deal with the day to day, with some guidance from MDG, and MDG could have a cherry picked version that was security audited for production
- Release in tandem with Node LTS versions releases: These versions are usually in preview for nearly a year before becoming the next LTS.

**Comment by MastroLindus** [comment]
@mrmowgli I think these points are a bit OT as it's more about concrete steps on where to take the future of meteor and less about the communication with MDG, but I'll answer quickly. I still think that both our posts should be masked from the main conversation though.

I would be against Meteor obliging you to work in a single, specific way, when supporting more is actually costing the same effort.
If somebody prefers React over Vue or vice-versa, I really don't see what's the benefit for meteor to restrict the use to one single case.
On the contrary, I can see many existing project maintainers that would get quite annoyed to having to re-change their code once again and re-learn again something new, when something that they already know works perfectly fine.
I am happy with Meteor making things 'easy' for me, without having to juggle between many configuration files and complex build steps. I am happy to have features like reactivity out of the box if I need it.
But if I cannot choose when or how to use something, or I cannot replace one part here or there, then the whole framework can become very quickly too limiting.

Imho the role of Meteor should be to allow you to reuse everything good that all the node/js community built and just make it really easy and straightforward to do so. Less of an implementation/framework provider, and more of super-easy assembling tool. 

I agree on your NPM point though. Atmosphere is past its time.

**Comment by sebakerckhof** [comment]
@mrmowgli Atmosphere is owned by MDG since they acqui-hired percolate studios some years back.

@KoenLav I like the idea of starting a github organization for popular packages. Myself and some others in this thread maintain much used packages or have outstanding PR's in popular packages that are no longer maintained. Bringing these all under one umbrella organization where maintainers can easily be added when somebody disappears seems like a good idea. But I feel that's not really relevant to the core of this discussion so maybe we should move it elsewhere (Or just create the organization and discuss it there)?

**Comment by mrmowgli** [comment]
@sebakerckhof I had forgotten about that, so I edited my comment.  Thanks for starting the community kickoff :+1: 

@MastroLindus I'm also ok with my posts being hidden from the main conversation. 
My comment around being opinionated isn't really about restricting what works with Meteor, but the narrative when people first come to the project.  It's hard to maintain great examples for beginners with just one framework, but now documentation and examples have to be made for Blaze, Angular and React.  Those examples become out of date, and need to track upstream new versions of Angular etc. It also leads to questions like: React is typically linked with GraphQL, so to use React in Meteor do I now need Apollo?   

To be clear, I'm not advocating removing supportor keeping people from adding their technology to Meteor, but rather MDG should settle on one proven stack to document around, officially support, and get people up to speed on.

**Comment by StorytellerCZ** [comment]
@sebakerckhof Me and @copleykj have made the first step towards community maintained packages: https://github.com/Meteor-Community-Packages
Once I get some more time I want to get there `alanning:roles` package (and release v2) and `bozhao:link-accounts`. Anyone else who wants to join in this effort is very much welcomed.

**Comment by KoenLav** [comment]
@sebakerckhof @StorytellerCZ @copleykj thanks for setting up the community organization!

I think the creation of this community, the invitation of several members and the collaboration within alone will help us join forces in maintaining Meteor packages.

For everyone who has previously forked Meteor packages to make them work in their own app, or is even maintaining some package for the community as a whole, please join the community by replying here: https://github.com/Meteor-Community-Packages/organization/issues/1

Also if you know some people who should be added definitely make sure that they are by mentioning them over there.

**Comment by KoenLav** [comment]
It's nice to see suggestions D (Welcoming community maintainers to Meteor) and I (Meteor Community Organization on GitHub) are already underway.

I would however still like to ask MDG, or more specifically @hwillson @abernix @benjamn, to chime in and share their thoughts on the suggestions raised above!

Could you possibly present us with an ETA for getting back on this? That would be much appreciated.

**Comment by KoenLav** [comment]
I've gone through the Meteor feature requests repository and noticed a lot of issues raised there have already been fixed or are no longer relevant.

I'd be happy to spend some time to clean up and organize that repository, in an effort to make it easier for the community to identify areas where contributions are encouraged/needed.

Still awaiting a reply from MDG.

**Comment by smeijer** [comment]
> Still awaiting a reply from MDG.

Don't hold your breath. This simply doesn't have their priority, and thereby I do not expect a reply at all.

They do reply when something is of their interest. Such as [Meteor Galaxy support](https://forums.meteor.com/t/ghosted-by-meteor-galaxy-support/49165).

I'm quite surprised by that topic; as it proves that MDG isn't dead. They just seem to be uninterested.

**Comment by smeijer** [comment]
https://github.com/meteor/meteor/issues/10477#issuecomment-468639980

> I wouldn't be surprised if Apollo would get some nice funding any time soon.

Well. There it is: https://twitter.com/GeoffQL/status/1138829967934910464?s=19

**Comment by arggh** [comment]
> I'm curious what this will mean for the meteor development.

Do you believe in Trickle-down economics...?

**Comment by trusktr** [comment]
For anyone above that mentioned TypeScript, the conversion of Meteor source code to TypeScript has already begun. Open up https://github.com/meteor/meteor/pull/10522 and search for "typescript". In particular, see Ben's announcement: https://github.com/meteor/meteor/pull/10522#issuecomment-508937874

**Comment by softwarecreations** [comment]
I'll add my 2 cents, will keep it short because this thread is very verbose already.

- I love Meteor, have been using it since 2012ish
- I'm very grateful to Members like Benjamin and the amazing development performance improvements we've gotten as well as React and NPM support.

- I whole-heartedly agree that Meteor has a communication problem.
- Meteor is an open source project but completely wastes the opportunity to leverage the community
- Meteor's marketing sucks
- Meteor appears to be dead on the surface. Yes, I know it's alive. I check the github, changelog and forum activity.

- Meteor's self sabotage in bad marketing kills it's popularity. That SO graph above showing it's decline is no joke.

Just think about the reality of this. Meteor appears to be dead, who will want to use it?
Who wants to pour their sweat blood and tears into technology that doesn't give them any confidence it'll be around or relevant in a year or two?
If you ask a developer how they know that (pick any project that's thriving) will be alive and relevant in a year or two, they'll tell you that the project's contributors and maintainers will ensure that it is. They'll add/update/fix whatever is needed to ensure it continues being great.

Why does Meteor appear to be dead/dying?
- Freenode IRC membership has declined from 250 down to 45 as of right now over the last few years. It's ABSOLUTELY dead. I've got ZNC running and I ask "Are you all bots?" and there's no reply for a couple of days. Good luck asking a question there!
- The basic parts of the guide (for example installing a hello world project, someone mentioned the accounts package hasn't been updated for 2 years) are badly outdated.
- Meteor lacks a social media person/team. Compare Meteor.com to the raspberrypi.org Meteor should have a person dedicated to improving the APPEARANCE, yes, marketing and communication of Meteor. It's not just lame stuff nerds, it's necessary and will bring more nerds. Yes I'm a nerd and used to not understand the relevance of marketing and communication either, but seriously, think about it.
- Runs on an ancient version of MongoDB. There doesn't appear be a clear indication of what versions of MongoDB Meteor supports.
- No official support for Preact (okay I'm not an expert, haven't tried Preact with Meteor, but Preact claims to do all the basic stuff react does at a tiny fraction of the bundle size, seems like it should be officially supported, can't be hard considering Meteor already supports React and NPM)

Please take the above as constructive criticism. I'd love to see Meteor succeed. It's still a special and relevant project in 2019. But a huge % of developers will turn their nose up at Meteor, for good reason. I understand that you guys are probably working on big, challenging things that are in demand, like typescript or whatever. But I feel this is probably poor prioritization. Meteor's #1 problem to solve is mindshare. Not even growing mindshare, but actually recovering all the mindshare they lost since 2015/6 would be a start. To gain mindshare we don't need new bells and whistles. We need to shine the old bells that we already have. We need to clean all the gunk out of them so they ring like new. Just sweep and vacuum the floor. Basic house keeping. It's a common problem that intelligent people have, they focus on the complicated problems when actually the simple problems will bring the greatest benefit with the least amount of time, money and energy spent.

**Comment by aliogaili** [comment]
I know what you are getting at but the hype you are after is not going to come back, Meteor passed that stage. The reality is that there are companies who made successful businesses out of Meteor (including MDG, qualia with 40M funding) and looking to maintaining and improving the technology behind it and I don't think they're interested in expanding the mindshare (MDG already pivoted to Apollo for revenue, although I believe Meteor + Apollo will be great ecosystem). Not all open source are that hyped, the issue with Meteor is that it raised the expectations so high during the hype peak with 41 million stars! The focus now on improving the underlaying build tool and decoupling the framework while supporting modern JS ecosystem, and I'm really happy with that direction and I'd rather see the effort focused there rather than growing more hype. Smart folks will see that and stick around, other will chase the hype.

**Comment by aliogaili** [comment]
Some other remarks on your comment:
> "Freenode IRC membership "

 I didn't even know that this channel existed, I think most of folks moved from IRC in general.

> "Runs on an ancient version of MongoDB."

 I'm not sure what you're referring to (I'm guessing the mongodriver) but I can easily bypass and use raw driver (which I rarely did).

> "No official support for Preact"
 
Meteor is positioned as front-end agnostic and the community can provide those integrations, plus preact is not that popular to have any official integration.

The fact that meteor is improving the underlying build tool and supporting typescript coupled with the fact that it's being used by successful startups makes it appear alive (from my perspective) so I don't even share your underlying premise that meteor is dying but surely the hype is dead, which is a good thing.

**Comment by softwarecreations** [comment]
> I don't even share your underlying premise that meteor is dying but surely the hype is dead, which is a good thing.
I never said Meteor is dying, please don't misquote me.
I said on the surface it appears to be dying, there's a lot of people who think it's dead/dying.
![image](https://user-images.githubusercontent.com/28704571/61549820-4a789480-aa51-11e9-8fad-fdc7ae69ff82.png)

Mindshare is very important. Mindshare is open source contributions and staying relevant. Mindshare is survival. Sure an open source project can drag on without mindshare, but why be on life support when you can thrive?

**Comment by aliogaili** [comment]
My sincere apology for misquoting you and yes unfortunately a lot of negative articles were written on Meteor 2016/17 declaring Meteor is dead, when in fact it got way better over the last few years. The x is dead titles tend to generate a lot of traffic and they're not restricted to Meteor (try is express dead or is ruby on rails dead). 

> Mindshare is survival.

I think funding is survival, some open source projects are too complex/specialized and there is a real need for dedicated resources to work on its core full-time. Without funding this is not possible, take a look at the [SemanticUI case](https://github.com/Semantic-Org/Semantic-UI/issues/6418). 

Anyway, for what I've seen Meteor has a really dedicated/loyal community and well funded companies behind it in addition to be being really great tech so I'm personally optimistic, but yeah things can improve.

**Comment by softwarecreations** [comment]
Yes, I'm mostly happy with meteor. I just think if they spent 1% of their mega millions worth of funding on marketing and keeping _basics_ updated it would bring the mindshare needed to massively improve Meteor in many areas.

Basics such as a router. The most commonly recommended router for Meteor is FlowRouter, which is 100% dead. This is bad. Makeup on a dead body only goes so far. It doesn't look as good as something that's alive.

**Comment by aliogaili** [comment]
> Why not spend 1% of those millions on basic things that matter to the common developer?

I guess because the common developer doesn't have/willing to pay the money to cover the 1% they would pay given how many JS frameworks out there, furthermore, the common JS developer is just looking for next framework on resume, better focus on enterprises. In my opinion, had they not pivoted to Apollo, MDG wouldn't be doing well today, a lot folks here are entrepreneurs and know how hard is it to build successful startups and it's even harder within open source.  But that is my guess, can't speak for MDG.

I personally would rather pay for better tooling, it's unfortunate most of those framework go only open source path.

**Comment by aliogaili** [comment]
Actually I want to add one more thing to this thread. 

With regards to mindshare, Benjamin has been working hard to make the code cleaner, more modular and modern over the last few years, take a look at [1.8.2 PR](https://github.com/meteor/meteor/pull/10522#issuecomment-511859094) and [these slide](https://slides.com/benjamn/reimplementing-meteor-in-typescripts). These are very clear signals of how committed MDG to supporting and engaging the community with Meteor going forward.  I really believe this is the right approach, asking MDG to put money in marketing is not but ensuring the code is more modular, modern, clean and well documented is.

Given Ben efforts in engaging the community, I think the community needs to step up and help with the typescript conversation and enhance the outdated documentation and we can close this issue.

**Comment by StorytellerCZ** [comment]
I agree with @aliogaili on most points, though I think there are some easy wins available to us outside of the ongoing tooling and typescript conversion (which has generated a lot of activity).
My personal favorite is updating the React packages to take advantage of new stuff in React (hooks). There are already PRs for that, we just need somebody with the privileges to jump in there, review, merge and publish.

**Comment by smeijer** [comment]
>  think the community needs to step up and help

It's not that the community isn't trying. See for example this topic, where some of the best and/or most active meteor contributors are trying to get in touch with MDG. But the problem is exactly in "engaging the community". There is no communication.

You say it's unfair to advise MDG to invest a bit in marketing and/or communication. But to be honest, they received 31 million funding for Meteor, and now 22 million for Apollo.

In the tweet, they repeat what they have said in the earlier funding rounds as well: _"Big thanks to the amazing Apollo community, we couldn't have done it without you!"_, and another quote: _"put every dollar toward our goal of helping app developers help the world."_. I believe communication and engaging with the community is very important to achieve that last goal.

So there is invested 53 million dollars in the Meteor and Apollo repo's. Is it really such a bad idea to get some marketing on this? How much would a community manager earn? 100k a year? Please give this person a contract for the next 5 years.

**Comment by aliogaili** [comment]
The gist of many of comments here is that MDG got another round investment and it should invest in hyping meteor again. I say hyping and not communication because just less than a week Ben, the core meteor developer put all his thoughts in these[slides](https://slides.com/benjamn/reimplementing-meteor-in-typescript#/) and it is as transparent as it gets.

My points:

1. Meteor is actually doing good, forum is active, packages are being created and core is getting better and better.
2. I disagree with using investors money for marketing, I'd rather see MDG keep pushing on the core build system and keep making the code more modular/cleaner. 
3. From what I've seen, open source longevity is better measured by community donations, loyalty and sponsors companies who actually use the tech in real businesses and not by investors money or hype, RethinkDB is one example that comes to mind.

Finally I think the [organization idea](https://github.com/Meteor-Community-Packages/organization/issues/1) is a great initiative that came out of this thread. Right now, a lot of great community packages are being created but they're not organized, I think what would be really good is to take that community organization idea and push it step further with a website/app that list the core meteor community packages,  a voting system for next features (similar to [webpack feature voting](https://webpack.js.org/vote/)) so it actually unify and organize the community voice. I think can build a simple meteor app (or maybe there is something existing for open source? ) that would initially do the following:
1. List core community members (people/or companies)
2. Allow voting on next roadmap items )with influencer system
3. Accept donation (that impact the influence points)
3. List core community packages 
4. List recommended community resources scatter around the internet 
The app/organization should be lean and easy to maintain going forward (meaning no need for for podcasts, or anything that takes a lot of manual time) since most of the folks on it will volunteer their time.

At least this would organize the voice/effort of the community.

**Comment by StorytellerCZ** [comment]
@aliogaili I'm just working on a static placeholder website that would fulfill the role of simple index page of all the activities. Nothing fancy. I'm down for more advance website. Let's hash out the ideas here:
https://github.com/Meteor-Community-Packages/organization/issues/5

**Comment by KoenLav** [comment]
@aliogaili I'm not for closing this thread until there is something of a proper response to some of the constructive criticisms raised here.

It has been almost 5 months since this thread was started. I know it's a lot to read, but there's a lot to improve. Right now it just looks like MDG doesn't care (enough).

**Comment by RobinJiao** [comment]
Obviously, people here all love Meteor, or they wouldn't waste their time writing these long posts; 
from a beginner's view, i can explain why people keep asking " is Meteor dead? "? I have the same feeling/worry.
I started to learn Meteor in Jan 2019, followed a youtube video and the official tutorial, managed to get my first app working,  and works well. then build 2-3 more tiny apps. really appreciate the Meteor framework. 
But it's really hard for a beginner:
1. almost all books are 3+ years old/outdated;
2. almost all video tutorials are 3+ old/outdated /almost all websites about meteor are gone
3. the questions about meteor in stackoverflow are often 4+ years old 
4. the necessary packages useraccounts/flowrouter/mup have very little commits for 2-3 years
5. Kadira team and many other developers left Meteor
seems nobody is using meteor anymore!
if meteor gets less new users(developers), the market gets smaller. people pay less attention to it. 
in brief, all we need is confidence - show us, please.

BTW, what people use after leaving Meteor, MERN? react+webpack +graphql? thanks
Robin Jiao

**Comment by Twisterking** [comment]
I will also give my 2 cents here, and somehow I can only pretty much agree with almost everyone on here.
I have been in love with Meteor pretty much from the first moment on, some time in 2012.
And it really worries me and almost hurts me to see this great, maybe best, fullstack JS framework out there making this very stupid mistake.

I agree, that there doesn't have to be a huch hype anymore like in the first 2 or 3 years. I get that. But as many said: **Meteor seems to be dead** (emphasis on the "seems" in this sentence)!

This is a huge problem indeed. Everyone who doesn't know Meteor for years now are afraid to use it, because it seems to be dead.

* Communication with the community is absolutely horrid
* Pretty much everything you can find is totally outdated (books, stackoverflow-posts, howtos, tutorials, guides, ... you name it)
* There *seems* to be no vision behind Meteor any more - The Roadmap and all that are also pretty much dead
* It *seems* like community contribution is either unwanted or MDG is not spending any resources to review and merge any PRs and stuff like that.

so **tl;dr**:
Meteor is still an awesome framework, but **please MDG, for the love of god: Don't let Meteor die and also don't let Meteor seem to be dead for that matter**! I don't agree with the people who say that things like this don't matter - they do matter. If enough people think you are dead, you are dead!

**Comment by smeijer** [comment]
> * Pretty much everything you can find is totally outdated (books, stackoverflow-posts, howtos, tutorials, guides, ... you name it)

This is not the first time I read this. I've created a topic on the forums for this subject, as I believe this can be easily fixed by the community. 

https://forums.meteor.com/t/new-meteor-book/49950

The question I ask there is:

> What are the questions you have? For what scope or for which subject do you wish to see tutorials / documentation? What is it that https://docs.meteor.com/ is lacking?

I'm willing to invest my time to deliver some content, but I need some content-requests to see what exactly is needed. As I myself, can manage fine with the docs.

**Comment by filipenevola** [comment]
Hi, I'm going to close this issue as MDG is not the owner of Meteor anymore.

I believe you already can see some communications from Meteor Software (the Tiny company behind Meteor) in the Forums, Twitter and Community Slack (you can join [here](https://join.slack.com/t/meteor-community/shared_invite/enQtODA0NTU2Nzk5MTA3LWY5NGMxMWRjZDgzYWMyMTEyYTQ3MTcwZmU2YjM5MTY3MjJkZjQ0NWRjOGZlYmIxZjFlYTA5Mjg4OTk3ODRiOTc) if you want).

Just today we published the new Meteor Roadmap https://github.com/meteor/meteor/blob/devel/Roadmap.md, this document was created based on community feedback.

Also, the stale bot is just to really close stale issues and not all the issues, `confirmed`, `pinned` and `security` issues will be preserved. We are going to review all of them and use the appropriated labels, as explained [here](https://github.com/meteor/meteor/blob/devel/ISSUE_TRIAGE.md).
